src/Symfony/Component/Messenger/Transport/Sender/Locator/AbstractSenderLocator.php

```diff
@@ -11,28 +11,26 @@
 
 namespace Symfony\Component\Messenger\Transport\Sender\Locator;
 
-use Symfony\Component\Messenger\Envelope;
-
 /**
  * @author Samuel Roze <samuel.roze@gmail.com>
  *
  * @internal
  */
 abstract class AbstractSenderLocator implements SenderLocatorInterface
 {
-    public static function getValueFromMessageRouting(array $mapping, Envelope $envelope)
+    public static function getValueFromMessageRouting(array $mapping, string $messageName)
     {
-        if (isset($mapping[$class = \get_class($envelope->getMessage())])) {
-            return $mapping[$class];
+        if (isset($mapping[$messageName])) {
+            return $mapping[$messageName];
         }
 
-        foreach (class_parents($class) as $name) {
+        foreach (class_parents($messageName) as $name) {
```

Comment:
since $messageName is not necessarily a class name anymore, this would fail then. You probably need to test class_exists